README
# Connie
Simple simulation of an electronic organ like e.g. [Vox Continental](http://en.wikipedia.org/wiki/Vox_Continental).

This program is intended as a toolbox for creating an electronic organ program with JACK MIDI input and JACK audio output for Linux. 
 
## Vox Continental
The *Vox Continental* transistorized combo organ creates the sound from twelve individual tunable LC oscillator boards with a six stage octave diver. The resulting 84 frequencies (RC filtered rectangular signals) are mixed with four contacts at key press and summed for the four stops. The different voices (flute and reed) are created by summing the original mix and a two stage RC filtered signal. The fixed switchable vibrato signal (RC filtered rectangular) introduces a fm at the oscillator boards.

For the *connie* program I looked at the schematics and wrote the software according to the signal flow with some tweaking of the voices. 

There is a lot of [info](http://www.reinout.nl/?page_id=7) and the [schematics](http://www.reinout.nl/?page_id=80) available for the Vox Continental.

## The Software
Connie is a JACK application. It has one MIDI input and a stereo audio output. The sound is generated by sampling lookup tables with calculated waves at  12 equal tempered frequencies for one octave. The upper octaves were generated by sampling with greater sample steps.

This solution allows a moderate cpu load.

 The sound of each note is ramped on and off to reduce the key click. For each sounding note four stops are mixed:
- The subharmonic 16' (one octave lower)
- The unison stop 8'
- The octave stop 4' (one octave higher)
- The mixture stop 2 2/3' (octave+fifth), 2' (superoctave), 
    1 3/5' (superoct.+major third), 1' (superoct.+octave)

To cover a broad variation of sounds three voices from two lookup tables can be mixed:
- Flute - a sine wave
- Reed - a bandlimited rectangle wave
- Sharp - a bandlimited sawtooth wave

For the reed/sharp voice I've created own samples of bandlimited signals for each octave. These samples are mixed at octave border to minimize the ugly "sound jump".  This combination allows a much wider sound range from very soft to very harsh. To get the typical cheesy "Connie" sound you have to fiddle with theese voices. I'll do more experiments with the waveforms and the lookup tables - but my first goal was to hear some noise ;)

Drawbars (vol = 0..8) control the four stops, the three voices and the intensity of the vibrato and percussion.

The vibrato works as a combined stero am and fm to create a sound like a very simple leslie - massive work has to be done, maybe with the help from some experts out there...

## Poor Man's Hammond
I'm working on a second test model with individual control of each stop and some other changes, kind of *poor-man's-hammond*, but (at the moment) without all the hammond special effects (leakage, click, random phase of wheels, vibrachorus etc.). 
This "hammond" project is just the extrapolation from the registers (16', 8', 4' and the mixture 2 2/3' + 2' + 1 3/5' + 1') already existing in connie to the well known nine numbers with a sine voice. This *p-m-h* model should therefore not be called *Hammond*, but *Test*.

Critics and hints are welcome, I need input for improvement. 

## Usage
    usage: connie [opts]
      -a                      autoconnect to system:playback ports
      -c CHANNEL              MIDI channel (1..16), 0=all (default)
      -f                      french AZERTY keyboard
      -g                      german QWERTZ keyboard
      -h                      this help msg
      -i INSTRUMENT           0: connie (default)
                              1: poor-man's-hammond
      -m MIDI_PORT            connect with midi port
      -p PITCH                concert pitch 220..880 Hz (default = 440 Hz)
      -s INTONATION_SCALE     0: Hammond Gears
                              1: Equally Tempered
                              2: Pure Intonation
                              3: Pythagorean
                              4: ¼ Comma Meantone
                              5: Werckmeister III
                              6: Kirnberger III
      -t TRANSPOSE            transpose -12..+12 semitones
      -v                      print version
      -C configfile           load config file
      -U UUID                 set jack session UUID


*VOX is a registered trademark of [VOX AMPLIFICATION LTD.](http://voxamps.com)*


Changelog
2020-10-30
connie 0.4.4 (time is on my side)
resurrected the project after 10 years sleep,
it still compiles and works under debian buster (stable) without any changes
fixed  "connie_main.c:658:29: warning: iteration 120 invokes undefined behavior [-Waggressive-loop-optimizations]"

2010-08-03
connie 0.4.3-rc6 (long time gone)
removed freeverb and added a JCRev style reverb (less cpu usage).
http://ccrma.stanford.edu/~jos/pasp/Schroeder_Reverberator_called_JCRev.html
DSP optimization, denormal handling
smoother attacK/decay envelopes
percussion triggered by first keypress only
diffent intonations
source code clean up
two binaries: connie (i386) and connie_sse (>= PIII with sse unit)
changed program options:
  -a			autoconnect to system:playback ports
  -c CHANNEL		MIDI channel (1..16), 0=all (default)
  -f			french AZERTY keyboard
  -g			german QWERTZ keyboard
  -h			this help msg
  -i INSTRUMENT		0: connie (default), 1: poor-man's-hammond
  -m MIDI_PORT		connect with midi port
  -p PITCH		concert pitch 220..880 Hz
  -s INTONATION_SCALE	 0: Hammond Gears
			 1: Equally Tempered
			 2: Pure Intonation
			 3: Pythagorean
			 4: ¼ Comma Meantone
			 5: Werckmeister III
			 6: Kirnberger III
  -t TRANSPOSE		transpose -12..+12 semitones
  -v			print version


2009-09-20
connie 0.4.2 (when the music's over)
MIDI in event handling at frame level instead of start of whole buffer.
No stuck notes anymore.
Experimental feature: freeverb
New cmd line arg:
  -c Chan : MIDI channel 1..16 or 0=all
  -p Freq : concert pitch 220..880 Hz

2009-06-24
connie 0.4.1 (summer in the city)
Created a second organ model (test): 9 stops, percussion and vibrato.
Cmd line args: 
  -a : autoconnect to system:playback ports
  -f : french AZERTY keyboard
  -g : german QWERTZ keyboard
  -t Num : organ type, 0 = connie (default), 1 = test
Modularized the source code.
Cleanup of user interface.

2009-06-13
connie 0.4.0 (land ho!)
I like this version - it's a big step forward
Last version "in the cage" used preconfigured mixture samples to 
reduce cpu load. The fixed freq ratio of 3:4:5:8 led to clean sound 
suitable for a church pipe organ but not for an analog multiplex organ.

This version combines the mixture from the equal temp. voices with
the effect of the vivid cutting sound.
New:
Three voices: flute=sine, reed=rect (sounds more like vox), sharp=saw
Percussion added
Leslie fm parameter reduced - sounds more realistic
Optimized realtime callback function reduces cpu load.
Midi cc 120 (all sound off) and 123 (all notes off) recognized
Midi pitch bend 
Cmd line argument --autoconnect

2009-06-09
connie 0.3.2 (in the cage)
name changed Connie -> connie
first debianized version
use different samples (only reed voice) for each octave
fill with bandlimited sawtooth (fmax < fsample/3) to avoid aliasing
average between sample buffer at octave border to avoid ugly sound "jumps"
vibrato parameter corrected
preset programs - select via pc kbd or midi

2009-06-07
Connie 0.3.1a (beautiful noise)
API changed: jack_client_new() -> jack_client_open()

2009-06-07
Connie 0.3.1 (beautiful noise)
note on/off with ramp to reduce key click
"gui" handling improved (set pc kbd non canonical)
better sounding default settings for drawbars
clean jack and term shutdown at ^C

2009-05-31
Connie 0.3 (house of the rising sun)
note switching at zero xing to reduce key clicking
reduced cpu load
adjustable vibrato

2009-05-30
Connie 0.2 (good vibrations)
added:
stereo output
vibrato does fm and pan shifting (simplified leslie sim)
Shift key -> panic

2009-05-29
First version: Connie 0.1 (light my fire)
features:
midi input noteon and noteoff and volume
audio output
text only "gui" in terminal
drawbar for stops and voices
fixed vibrato on/off


scales.h
/*****************************************************************************
 *
 *   scales.h
 *   Inspired by Fons' aeolus project
 *
 *   Simulation of an electronic organ like Vox Continental
 *   with JACK MIDI input and JACK audio output
 *
 *   Copyright (C) 2009,2010 Martin Homuth-Rosemann
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; version 2 of the License
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 ******************************************************************************/
 
#ifndef SCALES_H
#define SCALES_H

#define _GNU_SOURCE



// gear ratio of hammond organ
// http://de.wikipedia.org/wiki/Hammond-Orgel#Tonerzeugung
// motor speed 1200 rpm -> 20 revs/sec
// 2 teeth on lowest wheel -> 10 Hz * gear ratio
#define C (104.f / 85.f)
const float scale_hammond_gears[12] = { 
    C *  85.f / 104.f, // C (~8.2 Hz)
    C *  71.f /  82.f, // C#
    C *  67.f /  73.f, // D
    C * 105.f / 108.f, // D#
    C * 103.f / 100.f, // E
    C *  84.f /  77.f, // F
    C *  74.f /  64.f, // F#
    C *  98.f /  80.f, // G
    C *  96.f /  74.f, // G#
    C *  88.f /  64.f, // A (10 Hz * 88 / 64  = 13.75 Hz)
    C *  67.f /  46.f, // A#
    C * 108.f /  70.f  // B
};
#undef C



// Equal temperament
// step: 2 ** 1/12 = 1.059463094
// http://en.wikipedia.org/wiki/Equal_temperament
const float scale_equaltemp[12] =
{
    1.f,
    1.059463094f,
    1.122462048f,
    1.189207115f,
    1.259921050f,
    1.334839854f,
    1.414213562f,
    1.498307077f,
    1.587401052f,
    1.681792830f,
    1.781797436f,
    1.887748625f,
};



// Reine Stimmung (pure intonation)
// http://de.wikipedia.org/wiki/Reine_Stimmung#Erweiterte_reine_Stimmung
const float scale_pure[12] =
{
    1.f,
    16.f / 15.f,
    9.f / 8.f,
    6.f / 5.f,
    5.f / 4.f,
    4.f / 3.f,
    45.f / 32.f,
    3.f / 2.f,
    8.f / 5.f,
    5.f / 3.f,
    16.f / 9.f,
    15.f / 8.f,
};



// Pythagorean tuning
// http://en.wikipedia.org/wiki/Pythagorean_tuning
const float scale_pythagorean[12] =
{
    1.0f,
    256.f / 243.f,
    9.f / 8.f,
    32.f / 27.f,
    81.f / 64.f,
    4.f / 3.f,
    729.f / 512.f,
    3.f / 2.f,
    128.f / 81.f,
    27.f / 16.f,
    16.f / 9.f,
    243.f / 128.f,
};



// Quarter comma meantone (Pietro Aaron, 1523)
// http://en.wikipedia.org/wiki/Quarter-comma_meantone

#define X 1.044906727f
#define S 1.069984488f
#define T 1.118033989f
#define P 1.495348781f

const float scale_meanquart[12] = 
{
    1.f,
    X,
    T,
    T * S,
    T * T,
    T * T * S,
    T * T * T,
    P,
    P * X,
    P * T,
    P * T * S,
    P * T * T
};
#undef X 
#undef S
#undef T
#undef P



// Werckmeister temperament III (Andreas Werckmeister, 1681)
// http://en.wikipedia.org/wiki/Werckmeister_temperament

#define R2 1.414213562
#define RR2 1.189207115
#define RR8 1.681792831
const float scale_werckmeister3[12] = 
{
    1.f,
    256.f / 243.f,
    R2 * 64.f / 81.f,
    32.f / 27.f,
    RR2 * 256.f / 243.f,
    4.f / 3.f,
    1024.f / 729.f,
    RR8 * 8.f / 9.f,
    128.f / 81.f,
    RR2 * 1024.f / 729.f,
    16.f / 9.f,
    RR2 * 128.f / 81.f
};
#undef R2
#undef RR2
#undef RR8



// Kirnberger III (Johann Philipp Kirnberger, 1779?)
// http://groenewald-berlin.de/ttg/TTG_T093.html
// http://groenewald-berlin.de/tabellen/TAB-093.html

#define R5 2.236067977
#define RR5 1.495348781
#define RR125 3.343701525
const float scale_kirnberger3[12] = 
{
    1.0f,
    256.f / 243.f,
    R5 / 2.f,
    32.f / 27.f,
    5.f / 4.f,
    4.f / 3.f,
    45.f / 32.f,
    RR5,
    128.f / 81.f,
    RR125 / 2.f,
    16.f / 9.f,
    15.f / 8.f,
};
#undef R5
#undef RR5
#undef RR125



struct temper {
    const char *label;
    const float *f_ratio;
} scales[] =
{
    { "Hammond Gears",    scale_hammond_gears },
    { "Equally Tempered", scale_equaltemp },
    { "Pure Intonation",  scale_pure },
    { "Pythagorean",      scale_pythagorean },
    { "¼ Comma Meantone", scale_meanquart },
    { "Werckmeister III", scale_werckmeister3 },
    { "Kirnberger III",   scale_kirnberger3 },
};

const int NSCALES = sizeof( scales ) / sizeof( struct temper );

#endif

